datafusion/expr/src/binary_rule.rs

```diff
@@ -801,4 +804,230 @@ mod tests {
+            DataType::Utf8,
+            DataType::Utf8,
+            Operator::RegexMatch,
+            DataType::Utf8
+        );
+        test_coercion_binary_rule!(
+            DataType::Utf8,
+            DataType::Utf8,
+            Operator::RegexNotMatch,
+            DataType::Utf8
+        );
+        test_coercion_binary_rule!(
+            DataType::Utf8,
+            DataType::Utf8,
+            Operator::RegexNotIMatch,
+            DataType::Utf8
+        );
+        test_coercion_binary_rule!(
+            DataType::Int16,
+            DataType::Int64,
+            Operator::BitwiseAnd,
+            DataType::Int64
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_type_coercion_arithmetic() -> Result<()> {
+        // integer
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::UInt32,
+            Operator::Plus,
+            DataType::Int32
+        );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::UInt16,
+            Operator::Minus,
+            DataType::Int32
+        );
+        test_coercion_binary_rule!(
+            DataType::Int8,
+            DataType::Int64,
+            Operator::Multiply,
+            DataType::Int64
+        );
+        // float
+        test_coercion_binary_rule!(
+            DataType::Float32,
+            DataType::Int32,
+            Operator::Plus,
+            DataType::Float32
+        );
+        test_coercion_binary_rule!(
+            DataType::Float32,
+            DataType::Float64,
+            Operator::Multiply,
+            DataType::Float64
+        );
+        // decimal
+        // bug: https://github.com/apache/arrow-datafusion/issues/3387 will be fixed in the next pr
+        // test_coercion_binary_rule!(
+        //     DataType::Decimal128(10, 2),
+        //     DataType::Decimal128(10, 2),
+        //     Operator::Plus,
+        //     DataType::Decimal128(11, 2)
+        // );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::Decimal128(10, 2),
+            Operator::Plus,
+            DataType::Decimal128(13, 2)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::Decimal128(10, 2),
+            Operator::Minus,
+            DataType::Decimal128(13, 2)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::Decimal128(10, 2),
+            Operator::Multiply,
+            DataType::Decimal128(21, 2)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::Decimal128(10, 2),
+            Operator::Divide,
+            DataType::Decimal128(23, 11)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::Decimal128(10, 2),
+            Operator::Modulo,
+            DataType::Decimal128(10, 2)
+        );
+        // TODO add other data type
```

Comment:
I wonder if we should file another ticket to track this gap? Specifically, a "help wanted" ticket that explained what was needed might encourage some additional contributions